`Validate.IParamCollection.forAll`

omaus opened the pull request:
This PR
- adds the `forAll` function in `Validate.IParamCollection` module/type
- also adds an `ErrorMessage.ofIParamCollection` function according to this
  - provides a unit test for this functionality

omaus committed:
Add ErrorMessage for IParam collections

omaus committed:
Add `forAll` function to validate IParam collections

Closes #72.

omaus committed:
Add unit test for ErrorMessage (WIP)

omaus committed:
Add unit test for new ErrorMessage function

kMutagene commented on src/ARCExpect/ErrorMessage.fs:
```diff
@@ -85,6 +85,45 @@ type ErrorMessage =
+
+    static member ofIParamCollection error iParamCollection =
+
+        let iParam = Seq.head iParamCollection
```
I do not get these functions to be honest. The API naming indicates that it generates error messages from `IParam` collections, but you only use the first item in the sequence to generate the error message

omaus replied:
Yes, since printing all items would be too much. I mean, this is arbitrary. How many items do we want to display? I chose 1, but we could also do 3 or 5 or any number...
What's your opinion on this?

kMutagene replied:
> I mean, this is arbitrary.

Not really, ideally you want to know all items that do not satisfy the predicate

kMutagene commented on src/ARCExpect/ValidationFunctions.fs:
```diff
@@ -152,6 +152,18 @@ module Validate =
+            match Seq.forall projection paramCollection with
+            | true  -> ()
+            | false ->
+                ErrorMessage.ofIParamCollection $"does not exist" paramCollection
```
"does not exist" is not the correct error message here i think. since this should check wether the projection is true for all values, the same message as `ValueSatisfiesPredicate` ("is invalid) makes more sense.

However, there are 2 more issues here:
- please rename to `ValuesSatisfyPredicate`, as that is the name of the respective function for a single `IParam`. That way, it becomes `Validate.ParamCollection .ValuesSatisfyPredicate` ("Validate (for a) ParamCollection (that the) values satisfy (the) predicate")
- the error message seems to make even less sense to me combined with the new `ErrorMessage.ofIParamCollection`, since that uses the head of the sequence. Imagine the second value does not satisfy the predicate - but the error message uses the first - this is very confusing.

omaus replied:
It's just not about the Value but about the whole IParam. So the naming should be `ParamsSatisfyPredicate`.

To the second point: What would you think about reformulating the function so that it does not use `Seq.forAll` but iterates through the seq so that we can catch the failing IParams and return them?

kMutagene replied:
> It's just not about the Value but about the whole IParam

Ideally we would have all 3 variants for `Param` and `ParamCollection`.

Thinking again about this, i guess the names of the colelction versions should be a little different.


My suggestions would be

- `Param.SatisfiesPredicate`
- `Param.TermSatisfiesPredicate`
- `Param.ValueSatisfiesPredicate`
- `ParamCollection.ItemsSatisfyPredicate`
- `Param.TermsSatisfyPredicate`
- `Param.ValueySatisfyPredicate`

kMutagene commented:
@omaus how do you intend to proceed on this? for reference, the new API design guidelines are here: https://nfdi4plants.github.io/arc-validate/ARCExpect/design.html. 

Do you intend to start with a fresh PR or adapt the code in this one?

omaus commented:
> @omaus how do you intend to proceed on this? for reference, the new API design guidelines are here: https://nfdi4plants.github.io/arc-validate/ARCExpect/design.html.
> 
> Do you intend to start with a fresh PR or adapt the code in this one?

This is just about renaming, isn't it? So it should be much easier to just rename the functions instead of writing everything anew.

omaus committed:
Rename function to better match naming design

See: https://nfdi4plants.github.io/arc-validate/ARCExpect/design.html

omaus committed:
Update function to call first predicate failing param

omaus commented:
I updated the function's name to fit the naming design and also changed it in a way that the first param that does not satisfy the predicate is shown instead of a part of the whole collection.

kMutagene commented on src/ARCExpect/ValidationFunctions.fs:
```diff
@@ -152,6 +152,23 @@ module Validate =
+        /// </summary>
+        /// <param name="projection">A function that evaluates to true if the element satisfies the requirements.</param>
+        /// <param name="paramCollection">The IParam collection to validate.</param>
+        static member ParamsSatisfyPredicate (predicate : #IParam -> bool) (paramCollection : #seq<#IParam>) =
```
(also xml docs and function parameters do not match)

omaus replied:
Done. 👍

omaus committed:
Rename function according to naming design

See: https://nfdi4plants.github.io/arc-validate/ARCExpect/design.html

omaus committed:
Update XML tags according to renaming

omaus commented:
> and also changed it in a way that the first param that does not satisfy the predicate is shown instead of a part of the whole collection.

Btw. is this okay or would you prefer to show all predicate failing items?

omaus committed:
Let function failTest all unsatisfying items

omaus commented:
> > and also changed it in a way that the first param that does not satisfy the predicate is shown instead of a part of the whole collection.
> 
> Btw. is this okay or would you prefer to show all predicate failing items?

Okay, tweaked it so that it goes throw the whole collection.

kMutagene commented on src/ARCExpect/ValidationFunctions.fs:
```diff
@@ -165,6 +165,7 @@ module Validate =
                     if predicate en.Current |> not then
                         ErrorMessage.ofIParam $"does not satisfy predicate" en.Current
                         |> Expecto.Tests.failtestNoStackf "%s"
+                        loop ()
```
Have you tested wether this continues after 1 `failtestNoStackf` is executed? If that throws an exception i am quite sure that it wont

omaus replied:
You are correct. It is more complicated to do this than I thought...
But I think I'll get a solution.

omaus committed:
Update function to display all failing items at once

omaus committed:
Delete missed line

omaus commented on src/ARCExpect/ErrorMessage.fs:
```diff
@@ -85,6 +85,45 @@ type ErrorMessage =
+        str.ToString()
+
+
+    static member ofIParamCollection error iParamCollection =
```
@kMutagene
